File: grix-update/scripts/grix_update.py
#!/usr/bin/env python3
"""Update the grix-hermes skill bundle for Hermes deployments."""

from __future__ import annotations

import argparse
import json
import os
from pathlib import Path
import subprocess
import sys
from typing import Any


def clean_text(value: Any) -> str:
    return str(value or "").strip()


def default_install_dir() -> Path:
    hermes_home = Path(os.path.expanduser(clean_text(os.environ.get("HERMES_HOME")) or "~/.hermes")).resolve()
    return hermes_home / "skills" / "grix-hermes"


def project_root() -> Path:
    return Path(__file__).resolve().parents[2]


def run_command(cmd: list[str], *, cwd: str | None = None, check: bool = True) -> dict[str, Any]:
    result = subprocess.run(cmd, text=True, capture_output=True, cwd=cwd)
    payload = {
        "cmd": cmd,
        "cwd": cwd or "",
        "code": result.returncode,
        "stdout": (result.stdout or "").strip(),
        "stderr": (result.stderr or "").strip(),
    }
    if check and result.returncode != 0:
        raise RuntimeError(payload["stderr"] or payload["stdout"] or f"command failed: {' '.join(cmd)}")
    return payload


def detect_repo_state(repo_root: Path, git_cmd: str) -> dict[str, Any]:
    state = {
        "repo_root": str(repo_root),
        "has_git": (repo_root / ".git").exists(),
        "branch": "",
        "upstream": "",
        "dirty": False,
        "dirty_entries": [],
    }
    if not state["has_git"]:
        return state

    branch = run_command([git_cmd, "rev-parse", "--abbrev-ref", "HEAD"], cwd=str(repo_root))
    state["branch"] = branch["stdout"]
    status = run_command([git_cmd, "status", "--short"], cwd=str(repo_root))
    state["dirty_entries"] = [line for line in status["stdout"].splitlines() if line.strip()]
    state["dirty"] = bool(state["dirty_entries"])
    upstream = run_command(
        [git_cmd, "rev-parse", "--abbrev-ref", "--symbolic-full-name", "@{upstream}"],
        cwd=str(repo_root),
        check=False,
    )
    if upstream["code"] == 0:
        state["upstream"] = upstream["stdout"]
    return state


def build_plan(args: argparse.Namespace) -> dict[str, Any]:
    repo_root = Path(clean_text(args.repo_root) or project_root()).resolve()
    raw_install_dir = clean_text(args.install_dir)
    install_dir = Path(raw_install_dir).resolve() if raw_install_dir else None
    if install_dir is None and args.mode in {"apply-update", "check-and-apply"}:
        install_dir = default_install_dir()
    repo_state = detect_repo_state(repo_root, args.git)
    strategy = "git-pull"
    if not repo_state["has_git"]:
        if args.mode != "check-only":
            raise RuntimeError("Repository root is not a git checkout. Pass a git repo root for Hermes skill updates.")
        strategy = "inspect-only"

    commands: list[dict[str, Any]] = []

    if repo_state["has_git"]:
        commands.extend(
            [
                {"cmd": [args.git, "status", "--short"], "cwd": str(repo_root), "stage": "inspect"},
                {"cmd": [args.git, "rev-parse", "--abbrev-ref", "HEAD"], "cwd": str(repo_root), "stage": "inspect"},
                {"cmd": [args.git, "rev-parse", "--abbrev-ref", "--symbolic-full-name", "@{upstream}"], "cwd": str(repo_root), "stage": "inspect", "check": False},
            ]
        )
        if args.mode in {"check-only", "check-and-apply"}:
            commands.append({"cmd": [args.git, "fetch", "--prune"], "cwd": str(repo_root), "stage": "inspect"})
            if repo_state["upstream"]:
                commands.append(
                    {
                        "cmd": [args.git, "rev-list", "--left-right", "--count", f"HEAD...{repo_state['upstream']}"],
                        "cwd": str(repo_root),
                        "stage": "inspect",
                    }
                )
        if args.mode in {"apply-update", "check-and-apply"}:
            commands.append({"cmd": [args.git, "pull", "--ff-only"], "cwd": str(repo_root), "stage": "apply"})
            commands.append({"cmd": [args.npm, "install"], "cwd": str(repo_root), "stage": "apply"})

    if install_dir and args.mode in {"apply-update", "check-and-apply"}:
        commands.append(
            {
                "cmd": [args.node, str(repo_root / "bin" / "grix-hermes.mjs"), "install", "--dest", str(install_dir), "--force"],
                "cwd": str(repo_root),
                "stage": "apply",
            }
        )

    cron_ready = repo_state["has_git"] and not repo_state["dirty"]

    return {
        "repo_root": str(repo_root),
        "install_dir": str(install_dir) if install_dir else "",
        "mode": args.mode,
        "strategy": strategy,
        "allow_dirty": args.allow_dirty,
        "repo_state": repo_state,
        "cron_ready": cron_ready,
        "commands": commands,
    }


def main() -> int:
    parser = argparse.ArgumentParser(description="Update the grix-hermes skill bundle for Hermes.")
    parser.add_argument("--mode", choices=["check-only", "apply-update", "check-and-apply"], default="check-and-apply")
    parser.add_argument("--repo-root", default="")
    parser.add_argument("--install-dir", default="", help="Optional Hermes install dir to refresh after pulling the repo.")
    parser.add_argument("--allow-dirty", default="false", choices=["true", "false"])
    parser.add_argument("--git", default="git")
    parser.add_argument("--npm", default="npm")
    parser.add_argument("--node", default="node")
    parser.add_argument("--dry-run", action="store_true")
    parser.add_argument("--json", action="store_true")
    args = parser.parse_args()
    args.allow_dirty = args.allow_dirty == "true"

    try:
        plan = build_plan(args)
        results: list[dict[str, Any]] = []

        if not args.dry_run:
            repo_state = plan["repo_state"]
            if plan["mode"] in {"apply-update", "check-and-apply"} and repo_state["has_git"] and repo_state["dirty"] and not args.allow_dirty:
                raise RuntimeError("Repository has uncommitted changes; refuse to auto-update without --allow-dirty true.")

            for entry in plan["commands"]:
                cmd = entry["cmd"]
                cwd = entry["cwd"] or None
                check = entry.get("check", True)
                results.append(run_command(cmd, cwd=cwd, check=check))

        payload = {
            "ok": True,
            "dry_run": bool(args.dry_run),
            "results": results,
            **plan,
        }
        if args.json:
            print(json.dumps(payload, ensure_ascii=False, indent=2))
        else:
            print(f"mode={plan['mode']} strategy={plan['strategy']} dry_run={args.dry_run}")
            for entry in plan["commands"]:
                print("$ " + " ".join(entry["cmd"]))
        return 0
    except Exception as exc:  # noqa: BLE001
        payload = {"ok": False, "error": str(exc)}
        if args.json:
            print(json.dumps(payload, ensure_ascii=False, indent=2), file=sys.stderr)
        else:
            print(str(exc), file=sys.stderr)
        return 1


if __name__ == "__main__":
    raise SystemExit(main())

File: grix-update/scripts/grix_update.ts
#!/usr/bin/env node
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { spawnSync } from "node:child_process";
import { fileURLToPath } from "node:url";

type Mode = "check-only" | "apply-update" | "check-and-apply";

interface Flags {
  mode: Mode;
  repoRoot: string;
  installDir: string;
  allowDirty: boolean;
  git: string;
  npm: string;
  node: string;
  dryRun: boolean;
  json: boolean;
}

interface CommandEntry {
  cmd: string[];
  cwd: string;
  stage: "inspect" | "apply";
  check?: boolean;
}

interface CommandResult {
  cmd: string[];
  cwd: string;
  code: number;
  stdout: string;
  stderr: string;
}

interface RepoState {
  repo_root: string;
  has_git: boolean;
  branch: string;
  upstream: string;
  dirty: boolean;
  dirty_entries: string[];
}

interface Plan {
  repo_root: string;
  install_dir: string;
  mode: Mode;
  strategy: "git-pull" | "inspect-only";
  allow_dirty: boolean;
  repo_state: RepoState;
  cron_ready: boolean;
  commands: CommandEntry[];
}

function cleanText(value: unknown): string {
  return String(value ?? "").trim();
}

function expandHome(value: string): string {
  if (!value) return value;
  if (value === "~") return os.homedir();
  if (value.startsWith("~/")) return path.join(os.homedir(), value.slice(2));
  return value;
}

function projectRoot(): string {
  const here = path.dirname(fileURLToPath(import.meta.url));
  return path.resolve(here, "..", "..");
}

function defaultInstallDir(): string {
  const hermesHome = path.resolve(expandHome(cleanText(process.env.HERMES_HOME) || "~/.hermes"));
  return path.join(hermesHome, "skills", "grix-hermes");
}

function runCommand(cmd: string[], cwd: string | undefined, check = true): CommandResult {
  const [bin, ...rest] = cmd;
  if (!bin) throw new Error("runCommand received empty cmd");
  const result = spawnSync(bin, rest, { cwd: cwd || undefined, encoding: "utf8" });
  const payload: CommandResult = {
    cmd,
    cwd: cwd || "",
    code: result.status ?? -1,
    stdout: (result.stdout || "").trim(),
    stderr: (result.stderr || "").trim(),
  };
  if (check && payload.code !== 0) {
    throw new Error(payload.stderr || payload.stdout || `command failed: ${cmd.join(" ")}`);
  }
  return payload;
}

function detectRepoState(repoRoot: string, gitCmd: string): RepoState {
  const state: RepoState = {
    repo_root: repoRoot,
    has_git: fs.existsSync(path.join(repoRoot, ".git")),
    branch: "",
    upstream: "",
    dirty: false,
    dirty_entries: [],
  };
  if (!state.has_git) return state;

  const branch = runCommand([gitCmd, "rev-parse", "--abbrev-ref", "HEAD"], repoRoot);
  state.branch = branch.stdout;
  const status = runCommand([gitCmd, "status", "--short"], repoRoot);
  state.dirty_entries = status.stdout.split("\n").map((line) => line).filter((line) => line.trim());
  state.dirty = state.dirty_entries.length > 0;
  const upstream = runCommand(
    [gitCmd, "rev-parse", "--abbrev-ref", "--symbolic-full-name", "@{upstream}"],
    repoRoot,
    false,
  );
  if (upstream.code === 0) state.upstream = upstream.stdout;
  return state;
}

function buildPlan(flags: Flags): Plan {
  const repoRoot = path.resolve(cleanText(flags.repoRoot) || projectRoot());
  const rawInstallDir = cleanText(flags.installDir);
  let installDir = rawInstallDir ? path.resolve(expandHome(rawInstallDir)) : "";
  if (!installDir && (flags.mode === "apply-update" || flags.mode === "check-and-apply")) {
    installDir = defaultInstallDir();
  }
  const repoState = detectRepoState(repoRoot, flags.git);
  let strategy: Plan["strategy"] = "git-pull";
  if (!repoState.has_git) {
    if (flags.mode !== "check-only") {
      throw new Error(
        "Repository root is not a git checkout. Pass a git repo root for Hermes skill updates.",
      );
    }
    strategy = "inspect-only";
  }

  const commands: CommandEntry[] = [];
  if (repoState.has_git) {
    commands.push(
      { cmd: [flags.git, "status", "--short"], cwd: repoRoot, stage: "inspect" },
      { cmd: [flags.git, "rev-parse", "--abbrev-ref", "HEAD"], cwd: repoRoot, stage: "inspect" },
      {
        cmd: [flags.git, "rev-parse", "--abbrev-ref", "--symbolic-full-name", "@{upstream}"],
        cwd: repoRoot,
        stage: "inspect",
        check: false,
      },
    );
    if (flags.mode === "check-only" || flags.mode === "check-and-apply") {
      commands.push({ cmd: [flags.git, "fetch", "--prune"], cwd: repoRoot, stage: "inspect" });
      if (repoState.upstream) {
        commands.push({
          cmd: [flags.git, "rev-list", "--left-right", "--count", `HEAD...${repoState.upstream}`],
          cwd: repoRoot,
          stage: "inspect",
        });
      }
    }
    if (flags.mode === "apply-update" || flags.mode === "check-and-apply") {
      commands.push({ cmd: [flags.git, "pull", "--ff-only"], cwd: repoRoot, stage: "apply" });
      commands.push({ cmd: [flags.npm, "install"], cwd: repoRoot, stage: "apply" });
    }
  }

  if (installDir && (flags.mode === "apply-update" || flags.mode === "check-and-apply")) {
    commands.push({
      cmd: [
        flags.node,
        path.join(repoRoot, "bin", "grix-hermes.js"),
        "install",
        "--dest",
        installDir,
        "--force",
      ],
      cwd: repoRoot,
      stage: "apply",
    });
  }

  const cronReady = repoState.has_git && !repoState.dirty;

  return {
    repo_root: repoRoot,
    install_dir: installDir,
    mode: flags.mode,
    strategy,
    allow_dirty: flags.allowDirty,
    repo_state: repoState,
    cron_ready: cronReady,
    commands,
  };
}

function parseArgs(argv: string[]): Flags {
  const flags: Flags = {
    mode: "check-and-apply",
    repoRoot: "",
    installDir: "",
    allowDirty: false,
    git: "git",
    npm: "npm",
    node: "node",
    dryRun: false,
    json: false,
  };
  const allowedModes: Mode[] = ["check-only", "apply-update", "check-and-apply"];
  for (let i = 0; i < argv.length; i += 1) {
    const token = argv[i];
    const next = argv[i + 1];
    if (token === "--mode" && next) {
      if (!allowedModes.includes(next as Mode)) {
        throw new Error(`Invalid --mode: ${next}`);
      }
      flags.mode = next as Mode;
      i += 1;
      continue;
    }
    if (token === "--repo-root" && next !== undefined) {
      flags.repoRoot = next;
      i += 1;
      continue;
    }
    if (token === "--install-dir" && next !== undefined) {
      flags.installDir = next;
      i += 1;
      continue;
    }
    if (token === "--allow-dirty" && next !== undefined) {
      if (next !== "true" && next !== "false") {
        throw new Error(`Invalid --allow-dirty: ${next}`);
      }
      flags.allowDirty = next === "true";
      i += 1;
      continue;
    }
    if (token === "--git" && next !== undefined) {
      flags.git = next;
      i += 1;
      continue;
    }
    if (token === "--npm" && next !== undefined) {
      flags.npm = next;
      i += 1;
      continue;
    }
    if (token === "--node" && next !== undefined) {
      flags.node = next;
      i += 1;
      continue;
    }
    if (token === "--dry-run") {
      flags.dryRun = true;
      continue;
    }
    if (token === "--json") {
      flags.json = true;
      continue;
    }
  }
  return flags;
}

function main(): number {
  let flags: Flags;
  try {
    flags = parseArgs(process.argv.slice(2));
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    process.stderr.write(`${JSON.stringify({ ok: false, error: message })}\n`);
    return 1;
  }
  try {
    const plan = buildPlan(flags);
    const results: CommandResult[] = [];

    if (!flags.dryRun) {
      const repoState = plan.repo_state;
      if (
        (plan.mode === "apply-update" || plan.mode === "check-and-apply") &&
        repoState.has_git &&
        repoState.dirty &&
        !flags.allowDirty
      ) {
        throw new Error(
          "Repository has uncommitted changes; refuse to auto-update without --allow-dirty true.",
        );
      }
      for (const entry of plan.commands) {
        const check = entry.check !== false;
        results.push(runCommand(entry.cmd, entry.cwd || undefined, check));
      }
    }

    const payload = {
      ok: true as const,
      dry_run: flags.dryRun,
      results,
      ...plan,
    };
    if (flags.json) {
      process.stdout.write(`${JSON.stringify(payload, null, 2)}\n`);
    } else {
      process.stdout.write(
        `mode=${plan.mode} strategy=${plan.strategy} dry_run=${flags.dryRun}\n`,
      );
      for (const entry of plan.commands) {
        process.stdout.write(`$ ${entry.cmd.join(" ")}\n`);
      }
    }
    return 0;
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    const payload = { ok: false, error: message };
    if (flags.json) {
      process.stderr.write(`${JSON.stringify(payload, null, 2)}\n`);
    } else {
      process.stderr.write(`${message}\n`);
    }
    return 1;
  }
}

process.exit(main());
